### PR TITLE
Fakes support in TestPlatform

### PR DESCRIPTION
**Pull request by mayankbansal018**
Fixes https://github.com/Microsoft/vstest/issues/685

### CODE REVIEW

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs [hunk 1]
```diff
@@ -518,23 +518,18 @@ private void SendEvent(DataCollectionEventArgs args)
...
-                {
-                    EqtTrace.Verbose("DataCollectionManger:SendEvent: Raising event {0} to collector {1}", args.GetType(), dataCollectorInfo.DataCollectorConfig.FriendlyName);
-                }
 
-                try
-                {
-                    this.events.RaiseEvent(args);
-                }
-                catch (Exception ex)
+            //do not send events multiple times
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs]: :) Do we have test coverage for this?

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs [hunk 2]
```diff
@@ -518,23 +518,18 @@ private void SendEvent(DataCollectionEventArgs args)
...
+
+            try
+            {
+                this.events.RaiseEvent(args);
+            }
+            catch (Exception ex)
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs]: Is `catch` required? It seems we are already using `SafeInvoke` to raise the individual events, it captures and logs the exception?

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorInformation.cs [hunk 3]
```diff
@@ -176,7 +175,7 @@ internal void DisposeDataCollector()
         /// <summary>
         /// The get test execution environment variables sync.
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorInformation.cs]: nit: out of date comment

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorInformation.cs [hunk 4]
```diff
@@ -176,7 +175,7 @@ internal void DisposeDataCollector()
...
         /// </summary>
-        private void SetTestExecutionEnvironmentVariables()
+        public void SetTestExecutionEnvironmentVariables()
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorInformation.cs]: Why is this change required?

File: src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        {
+            try
+            {
+                var assemblyPath = Path.Combine(Path.GetDirectoryName(Assembly.GetEntryAssembly().Location), FakesConfiguratorAssembly + ".dll");
+
+                Assembly assembly = null;
+#if NET46
+                assembly = Assembly.LoadFrom(assemblyPath);
+#else
+                // I think we should not do this, since FakeConfigurator is built for net46?
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs]: Yes, we shouldn't do this.

Comment (codito) [inline src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs]: Possibly we can remove the System.Reflection reference in csproj?

File: src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+        private static bool TryGetFakesDataCollectorConfigurator(out Func<IEnumerable<string>, string> configurator)
+        {
+            try
+            {
+                var assemblyPath = Path.Combine(Path.GetDirectoryName(Assembly.GetEntryAssembly().Location), FakesConfiguratorAssembly + ".dll");
+
+                Assembly assembly = null;
+#if NET46
+                assembly = Assembly.LoadFrom(assemblyPath);
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs]: Is it okay to use `Assembly.Load` in this case since TP.Fakes.dll is always in the app base?

File: src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            Func<IEnumerable<string>, string> configurator;
+
+            // fakes supported?
+            if (!TryGetFakesDataCollectorConfigurator(out configurator))
+            {
+                return false;
+            }
+
+            // if no fakes, return settings unchanged
+            var fakesConfiguration = configurator(sources);
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs]: Can this configuration be done within the datacollector itself?

It will be difficult to perform such magic for external data collectors. May be our datacollector extensibility APIs require an addition to support this?

File: src/vstest.console/Processors/DisableAutoFakesArgumentProcessor.cs [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    using System;
+    using System.Globalization;
+
+    using CommandLineResources = Microsoft.VisualStudio.TestPlatform.CommandLine.Resources.Resources;
+
+    /// <summary>
+    /// An argument processor that allows the user to disable fakes
+    /// from the command line using the --DisableAutoFakes|/DisableAutoFakes command line switch.
+    /// </summary>
+    internal class DisableAutoFakesArgumentProcessor : IArgumentProcessor
```
Comment (codito) [inline src/vstest.console/Processors/DisableAutoFakesArgumentProcessor.cs]: This is interesting. For `code coverage` we require users to specify `/collect`, for fakes we're automatically trying to figure if fakes should be enabled from the provided test sources. I believe this is for compatibility with TPv1.

cc: @Faizan2304 @pvlakshm fyi

Comment (codito) [inline src/vstest.console/Processors/DisableAutoFakesArgumentProcessor.cs]: I think this will break cloud build if we don't auto enable fakes. Is it a possibility that this can be done in the caller of `vstest.console.exe`?

File: test/Microsoft.TestPlatform.Common.UnitTests/Utilities/FakesUtilitiesTests.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace TestPlatform.Common.UnitTests.Utilities
```
Comment (codito) [inline test/Microsoft.TestPlatform.Common.UnitTests/Utilities/FakesUtilitiesTests.cs]: We need positive tests?

Comment (codito) [inline test/Microsoft.TestPlatform.Common.UnitTests/Utilities/FakesUtilitiesTests.cs]: nit: should the namespace be fully qualified?

File: src/vstest.console/Processors/DisableAutoFakesArgumentProcessor.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    internal class DisableAutoFakesArgumentProcessorCapabilities : BaseArgumentProcessorCapabilities
+    {
+        public override bool AllowMultiple => false;
+        public override string CommandName => DisableAutoFakesArgumentProcessor.CommandName;
+
+        public override ArgumentProcessorPriority Priority => ArgumentProcessorPriority.Normal;
+
+        public override HelpContentPriority HelpPriority => HelpContentPriority.DisableAutoFakesArgumentProcessorHelpPriority;
```
Comment (codito) [inline src/vstest.console/Processors/DisableAutoFakesArgumentProcessor.cs]: Does this require help text?

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorInformation.cs (hunk 4 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectorInformation.cs]: We are delaying getting Environment variables for datacollectors. If we do it early(that is before initializing all DataCollectors) then it creates a partial collection plan, which contains info about only one(current) datacollector.

File: src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs (hunk 5 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs]: If we remove this then we can remove reference to System,Reflection, I did it for now so that if we do work for fakes in NetCore, we won't need to make change here.
Either way is fine with me.

File: src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs (hunk 7 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs]: The whole logic to check whether or not to enable fakes lies inside this function (configurator), what you are suggesting is correct, but that would need for us to refactor code so that atleast we figure out whether to enable fakes or not, & then generate appropriate settings for it inside it's data collector.

File: src/vstest.console/Processors/DisableAutoFakesArgumentProcessor.cs (hunk 10 above)
Comment (mayankbansal018) [inline src/vstest.console/Processors/DisableAutoFakesArgumentProcessor.cs]: It was present before, so I let it be

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs (hunk 2 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs]: yes I'll remove it

File: src/vstest.console/Processors/DisableAutoFakesArgumentProcessor.cs (hunk 10 above)
Comment (mayankbansal018) [inline src/vstest.console/Processors/DisableAutoFakesArgumentProcessor.cs]: The text is there, but is not displayed since the property HelpContentResourceName is not assigned

File: src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs (hunk 6 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Common/Utilities/FakesUtilities.cs]: Resolved
